cmd/kubeadm/app/phases/kubelet/config.go

```diff
@@ -30,16 +30,29 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 	"k8s.io/apimachinery/pkg/util/version"
 	clientset "k8s.io/client-go/kubernetes"
+	"k8s.io/klog"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
 	"k8s.io/kubernetes/cmd/kubeadm/app/componentconfigs"
 	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	kubeadmutil "k8s.io/kubernetes/cmd/kubeadm/app/util"
 	"k8s.io/kubernetes/cmd/kubeadm/app/util/apiclient"
 	rbachelper "k8s.io/kubernetes/pkg/apis/rbac/v1"
 	kubeletconfig "k8s.io/kubernetes/pkg/kubelet/apis/config"
+	utilsexec "k8s.io/utils/exec"
 )
 
 // WriteConfigToDisk writes the kubelet config object down to a file
 // Used at "kubeadm init" and "kubeadm upgrade" time
-func WriteConfigToDisk(kubeletConfig *kubeletconfig.KubeletConfiguration, kubeletDir string) error {
+func WriteConfigToDisk(kubeletConfig *kubeletconfig.KubeletConfiguration, nodeReg *kubeadmapi.NodeRegistrationOptions, kubeletDir string) error {
```

Comment:
kubeadm join does not use this function when writing the file to disk.
it uses:
https://github.com/kubernetes/kubernetes/blob/a404238d1c90ee00827624249b9223e7644e3f99/cmd/kubeadm/app/cmd/phases/join/kubelet.go#L134-L137

then it uses this private function:
`writeConfigBytesToDisk`
https://github.com/kubernetes/kubernetes/blob/9c75270b2be4dcc39e31edb34b76e8f604aaf91a/cmd/kubeadm/app/phases/kubelet/config.go#L157

so the driver detection on join will still result in a flag instead of a config value.

both init and join end up calling `writeConfigBytesToDisk`, but i don't think we should have the driver detection in there.

this needs a much larger refactor.
instead of moving only the cgroup driver detection away from `cmd/kubeadm/app/phases/kubelet/flags.go` my idea was to move all of the flags that can be under the config from here:
https://github.com/kubernetes/kubernetes/blob/e47fc04adc39c06bcdd6d7dc74ee881d8740a81a/cmd/kubeadm/app/phases/kubelet/flags.go#L73-L115

into a new intermediate step that applies to the KubeletConfiguration that is written to disk on both init and join.

Reply:
@neolit123 you are right. Currently there are a couple of flags in that code `cgroup-driver` and `resolv-conf` with the latter one being clearly host specific. We need to remove them from there and allow for host specific fields of the kubelet component config to be patched locally (either on create or on download of the config).